src/include/catalog/database_catalog.h

```diff
@@ -529,151 +263,85 @@ class DatabaseCatalog {
+  template <typename Column, typename ClassOid, typename ColOid>
+  bool CreateColumn(common::ManagedPointer<transaction::TransactionContext> txn, ClassOid class_oid, ColOid col_oid,
+                    const Column &col);
+  /** @brief Get all the columns for a particular pg_attribute entry. @see PgAttributeImpl::GetColumns */
+  template <typename Column, typename ClassOid, typename ColOid>
+  std::vector<Column> GetColumns(common::ManagedPointer<transaction::TransactionContext> txn, ClassOid class_oid);
+  /** @brief Delete all columns for a particular pg_attribute entry. @see PgAttributeImpl::DeleteColumns */
+  template <typename Column, typename ClassOid>
+  bool DeleteColumns(common::ManagedPointer<transaction::TransactionContext> txn, ClassOid class_oid);
 
   /**
-   * Sets a table's schema in pg_class
-   * @warning Should only be used by recovery
-   * @param txn transaction to query
-   * @param oid oid to object
-   * @param schema object schema to insert
-   * @return true if succesfull
+   * @brief Set the schema of a table in pg_class.
+   *
+   * @tparam CallerType     The type of the caller. Should only be used by recovery!
+   * @param txn             The transaction to perform the schema change in.
+   * @param oid             The OID of the table.
+   * @param schema          The new schema to set.
+   * @return                True if the schema was set successfully. False otherwise.
    */
+  template <typename CallerType>
   bool SetTableSchemaPointer(common::ManagedPointer<transaction::TransactionContext> txn, table_oid_t oid,
-                             const Schema *schema);
+                             const Schema *schema) {
+    static_assert(std::is_same_v<CallerType, storage::RecoveryManager>, "Only recovery should call this.");
+    return SetClassPointer(txn, oid, schema, postgres::PgClass::REL_SCHEMA.oid_);
+  }
 
   /**
-   * Sets an index's schema in pg_class
-   * @warning Should only be used by recovery
-   * @param txn transaction to query
-   * @param oid oid to object
-   * @param schema object schema to insert
-   * @return true if succesfull
+   * @brief Set the schema of an index in pg_class.
+   *
+   * @tparam CallerType     The type of the caller. Should only be used by recovery!
+   * @param txn             The transaction to perform the schema change in.
+   * @param oid             The OID of the index.
+   * @param schema          The new index schema to set.
+   * @return                True if the index schema was set successfully. False otherwise.
    */
+  template <typename CallerType>
   bool SetIndexSchemaPointer(common::ManagedPointer<transaction::TransactionContext> txn, index_oid_t oid,
-                             const IndexSchema *schema);
+                             const IndexSchema *schema) {
+    static_assert(std::is_same_v<CallerType, storage::RecoveryManager>, "Only recovery should call this.");
```

Comment:
This is slick!  This may deserve some documentation in the `docs` folder as a technique for enforcing caller constraints elsewhere in the codebase.  I think we have warnings about this in other places but lack actual enforcement mechanisms.

Reply:
Added a bit to `cpp_guidelines_code_style`, take a look? thanks!